src/librustc_mir/borrow_check/move_errors.rs

```diff
@@ -291,111 +294,119 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
+            GroupedMoveError::MovesFromValue { mut binds_to, .. } => {
                 binds_to.sort();
                 binds_to.dedup();
-                let mut multipart_suggestion = Vec::with_capacity(binds_to.len());
-                for (j, local) in binds_to.into_iter().enumerate() {
-                    let bind_to = &self.mir.local_decls[local];
-                    let binding_span = bind_to.source_info.span;
+                self.add_move_error_suggestions(err, &binds_to);
+                self.add_move_error_labels(err, &binds_to);
+            }
+            // No binding. Nothing to suggest.
+            GroupedMoveError::OtherIllegalMove { .. } => (),
+        }
+    }
 
-                    // Suggest ref mut when the user has already written mut.
-                    let ref_kind = match bind_to.mutability {
-                        Mutability::Not => "ref",
-                        Mutability::Mut => "ref mut",
-                    };
-                    if j == 0 {
-                        err.span_label(binding_span, format!("data moved here"));
+    fn add_move_error_suggestions(
+        &self,
+        err: &mut DiagnosticBuilder<'a>,
+        binds_to: &[Local],
+    ) {
+        for local in binds_to {
+            let bind_to = &self.mir.local_decls[*local];
+            if let Some(
+                ClearCrossCrate::Set(BindingForm::Var(VarBindingForm {
+                    pat_span,
+                    ..
+                }))
+            ) = bind_to.is_user_variable {
+                let pat_snippet = self
+                    .tcx.sess.codemap()
+                    .span_to_snippet(pat_span)
+                    .unwrap();
+                if pat_snippet.starts_with('&') {
+                    let pat_snippet = &pat_snippet[1..];
+                    let suggestion;
+                    if pat_snippet.starts_with("mut")
+                        && pat_snippet["mut".len()..].starts_with(Pattern_White_Space)
+                    {
+                        suggestion = pat_snippet["mut".len()..].trim_left();
                     } else {
-                        err.span_label(binding_span, format!("... and here"));
-                    }
-                    match bind_to.name {
-                        Some(name) => {
-                            multipart_suggestion.push((binding_span,
-                                                       format!("{} {}", ref_kind, name)));
-                        }
-                        None => {
-                            err.span_label(
-                                span,
-                                format!("Local {:?} is not suitable for ref", bind_to),
-                            );
-                        }
+                        suggestion = pat_snippet;
                     }
+                    err.span_suggestion(
+                        pat_span,
+                        "consider removing this borrow operator",
```

Comment:
Nit: perhaps "consider removing the `&`" / "consider removing the `&mut`" would be clearer that "consider removing this borrow operator"?

Reply:
IMO "consider removing this borrow operator: `Foo::Bar(..)`" looks confusing, since "this" makes it sound like it's referring to `Foo::Bar(..)` as a "borrow operator", which isn't what's going on. I personally find "consider removing the `&`: `Foo::Bar(...)`" easier to read.

Reply:
In the (probably rare) case where the pattern is a double reference (`&&Foo::Bar(..)`), "consider not borrowing" would be sorta imprecise. But that might not be worth worrying about.

Reply:
Although, now that I think about it, I guess we could strip *all* the leading `&`/`&mut` instead of just the first one.

Reply:
That is if we don't change the span either. If we do, this could look like this:

```
error[E0507]: cannot move out of borrowed content
  --> $DIR/dont-suggest-ref.rs:103:18
   |
LL |     let &X(_t) = s;
   |         -  --    ^ cannot move out of borrowed content
   |         |  |
   |         |  data moved here
   |         |  move occurs because `_t` has type `Y`, which does not implement the `Copy` trait
   |         help: consider removing this borrow
```